app/src/main/java/com/github/multimatum_team/multimatum/model/DeadlineAdapter.kt

```diff
@@ -42,24 +42,33 @@ class DeadlineAdapter(private val context: Context) : BaseAdapter() {
             .fromApplication(context, ClockServiceEntryPoint::class.java)
             .provideClockService()
 
-    private var dataSource: SortedMap<DeadlineID, Deadline> = sortedMapOf()
+    private var dataSource: List<Pair<DeadlineID, Deadline>> = listOf()
 
     private val inflater: LayoutInflater =
         context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
 
     fun setDeadlines(deadlines: Map<DeadlineID, Deadline>) {
-        dataSource = deadlines.toSortedMap { id1, id2 ->
-            compareValues(deadlines[id1]!!.dateTime, deadlines[id2]!!.dateTime)
-        }
+        dataSource = sortDeadline(deadlines)
+                // deadlines.toSortedMap { id1, id2 -> compareValues(deadlines[id1]!!.dateTime, deadlines[id2]!!.dateTime) }
         notifyDataSetChanged()
     }
 
+    private fun sortDeadline(deadlines: Map<DeadlineID, Deadline>): List<Pair<DeadlineID, Deadline>> {
+        val doneDeadlines: List<Pair<DeadlineID, Deadline>> = deadlines
```

Comment:
Instead of duplicating the code for each deadline state, consider doing something like:
```kt
    val partition: MutableMap<DeadlineState, MutableList<Pair<DeadlineID, Deadline>>> = DeadlineState
        .values()
        .associate { state -> Pair<DeadlineState, MutableList<Pair<DeadlineID, Deadline>>>(state, mutableListOf()) }
       	.toMutableMap()
      
    for ((id, deadline) in deadlines.entries) {
        partition[deadline.state]!!.add(Pair(id, deadline))
    }
    
    val result: MutableList<Pair<DeadlineID, Deadline>> = mutableListOf()
    for (state in DeadlineState.values()) {
        result.addAll(partition[state]!!.sortedBy { it.second.date })
    }
```
This way the code will continue working even if we add new deadline states.

Reply:
can you explain me briefly what this code do. I'm not sur to understand why we need to use mutableMap

Reply:
This snippet groups deadlines by states, then concatenates the groups sorted by dates. Doing so avoids traversing the list of deadlines twice, and will work even if we define new deadline states other than just todo and done.
You're right that we don't need the MutableMap for partition, an immutale Map is enough.